### PR TITLE
Extend STOP statement following GnuCOBOL syntax

### PR DESCRIPTION
**Pull request by engboris**
There was a lot of missing syntactic constructions for `STOP` which are defined in GnuCOBOL.

References:
- https://www.microfocus.com/documentation/visual-cobol/vc60/DevHub/HRLHLHPDFC06.html
- https://github.com/OCamlPro/gnucobol/blob/89c45a3bc80c2ca6302382bd40fdc585436d65f9/cobc/parser.y#L16731

### CODE REVIEW

File: src/lsp/cobol_parser/grammar.mly [hunk 1]
```diff
@@ -3998,17 +3998,38 @@ let start_position ==
...
+let stop_run_returning_body :=
+  | ~ = scalar; <StopReturnScalar> 
+  | or_(GIVING, RETURNING); o(ADDRESS; OF); ~ = qualident; <StopReturnAddress>
+  | or_(GIVING, RETURNING); v = integer;
+    { StopReturnInt { value = Integer v;
+                      size = None } }
+  | or_(GIVING, RETURNING); v = integer; s = pf(SIZE; IS, integer);
+    { StopReturnInt { value = Integer v;
+                      size = Some (Integer s) } }
+  | ~ = with_status; <StopReturnStatus>
```
Comment (ddeclerck) [inline src/lsp/cobol_parser/grammar.mly:4017]: Are all those in the standard ? Or are they mostly extensions ?

Comment (engboris) [reply src/lsp/cobol_parser/grammar.mly:4017]: There are a lot of extensions (common, RM/COBOL and GCOS according to GnuCOBOL). I did not find information on `| stop thread_reference_optional` and I did not implement it. I think I'll do it in this PR.

File: src/lsp/cobol_ptree/simple_statements.ml [hunk 2]
```diff
@@ -781,36 +781,70 @@ let pp_sort_stmt ppf = function
...
+  | StopRun of stop_run_body option
+  | StopError
 [@@deriving ord]
 
-and stop_run =
+and stop_arg =
+  | StopWithQualIdent of qualident
+  | StopWithLiteral of literal
+
+and stop_run_body =
```
Comment (nberth) [inline src/lsp/cobol_ptree/simple_statements.ml:793]: ~~`_body`~~/`_return`

File: src/lsp/cobol_ptree/simple_statements.ml [hunk 3]
```diff
@@ -781,36 +781,70 @@ let pp_sort_stmt ppf = function
...
+  | StopReturnScalar of scalar
```
Comment (nberth) [inline src/lsp/cobol_ptree/simple_statements.ml:794]: I'd go with `StopReturningScalar` (and same prefix for the other constructors).

File: src/lsp/cobol_ptree/simple_statements.ml [hunk 4]
```diff
@@ -781,36 +781,70 @@ let pp_sort_stmt ppf = function
...
+    pp_status_kind status_kind
+    Fmt.(option pp_scalar) status_value
 
 let pp_stop_stmt ppf = function
-  | StopRun sro ->
-    Fmt.pf ppf "STOP@ RUN%a"
-      Fmt.(option (sp ++ pp_stop_run)) sro
-  | StopLiteral lit -> Fmt.pf ppf "STOP@ %a" pp_literal lit
+  | StopArg None -> Fmt.pf ppf "STOP"
+  | StopArg (Some a) -> begin match a with
```
Comment (nberth) [inline src/lsp/cobol_ptree/simple_statements.ml:827]: Here you should just flatten the pattern:
```
| StopArg Some StopWithQualIdent i ->
| StopArg Some StopWithLiteral l ->
```
(same below)

**Review by nberth — COMMENTED** [review]
Nice additions!
Just minor comments on names and formatting.

**Comment by engboris** [comment]
I also added `STOP THREAD identifier?` as in GnuCOBOL

**Review by nberth — APPROVED** [review]

**Pull request merged**